packages/rpc/src/Client.ts

```diff
@@ -1,67 +1,115 @@
-import { assert } from '@chainflip/utils/assertion';
+import { type DeferredPromise, deferredPromise } from '@chainflip/utils/async';
 import {
   type RpcRequest,
   type RpcMethod,
   type RpcResult,
   rpcResult,
-  rpcResponse,
   type JsonRpcRequest,
+  type JsonRpcResponse,
 } from './common';
 
 export type Response =
-  | { success: true; id: string; result: unknown }
+  | { success: true; id: string; result: JsonRpcResponse }
   | { success: false; id: string; error: Error };
+
+export type RequestMap = Map<
+  string,
+  {
+    deferred: DeferredPromise<RpcResult<RpcMethod>>;
+    body: JsonRpcRequest<RpcMethod>;
+    method: RpcMethod;
+  }
+>;
+
 export default abstract class Client {
   private lastRequestId = 0;
+  private timer: ReturnType<typeof setTimeout> | null = null;
+  private requestMap: RequestMap = new Map();
 
   constructor(protected readonly url: string) {}
 
   protected abstract send<const T extends RpcMethod>(
     data: JsonRpcRequest<T>[],
-  ): Promise<Response[]>;
+    clonedMap: RequestMap,
+  ): Promise<void>;
 
-  protected getRequestId() {
-    return String(++this.lastRequestId);
+  private getRequestId() {
+    try {
+      return crypto.randomUUID();
+    } catch {
+      return String(++this.lastRequestId);
+    }
   }
 
-  protected formatRequest<T extends RpcMethod>(
-    method: T,
-    params: RpcRequest[T],
-  ): JsonRpcRequest<T> {
+  private formatRequest<T extends RpcMethod>(method: T, params: RpcRequest[T]): JsonRpcRequest<T> {
     return { jsonrpc: '2.0', id: this.getRequestId(), method, params } as const;
   }
 
-  protected parseSingleResponse(response: Response) {
+  private parseSingleResponse(response: Response): unknown {
     if (!response.success) {
       throw response.error;
     }
 
-    const parseResult = rpcResponse.safeParse(response.result);
+    const parseResult = response.result;
 
-    if (!parseResult.success) {
-      throw new Error('Malformed RPC response received');
+    if ('error' in parseResult) {
+      throw new Error(`RPC error [${parseResult.error.code}]: ${parseResult.error.message}`);
     }
 
-    if ('error' in parseResult.data) {
-      throw new Error(
-        `RPC error [${parseResult.data.error.code}]: ${parseResult.data.error.message}`,
-      );
+    return parseResult.result;
```

Comment:
okay im more confused now - `response.result` was already parsed with a zod schema then?

Reply:
okay no it wasnt - we pass the jsonrpc result object as `response.result`, which then has another `result` that contains the data

Reply:
there's layers here for sure